funsor/ops/op.py

```diff
@@ -1,9 +1,30 @@
 # Copyright Contributors to the Pyro project.
 # SPDX-License-Identifier: Apache-2.0
 
+import weakref
+
 from multipledispatch import Dispatcher
 
 
+class OpCacheMeta(type):
+    """
+    Metaclass for caching op instance construction.
+    """
+    # use a WeakKeyDictionary to allow garbage collection of out-of-scope op classes.
+    # this is especially important when some ops are dynamically generated and
+    # their instances are large, e.g. an op corresponding to a neural network.
+    _cls_caches = weakref.WeakKeyDictionary()  # class -> {args: instance}
```

Comment:
Replace this with a per-class dictionary that is created during `.__init_subclass__()`, thereby more closely mimicking existing behavior.
```py
def __init_subclass__(cls, default_name, **kwargs):
    super().__init_subclass__(**kwargs)
    cls._cache = {}
```